src/lib-imap/imap-capability-list.h

```diff
@@ -0,0 +1,85 @@
+#ifndef IMAP_CAPABILITY_LIST_H
+#define IMAP_CAPABILITY_LIST_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "lib.h"
+#include "array.h"
+
+/* bitflags to represent different conditions of
+   the current imap session */
+enum imap_capability_visibility {
+	IMAP_CAP_VISIBILITY_NEVER           = 0,      /* never show this capability */
+	IMAP_CAP_VISIBILITY_PREAUTH         = BIT(0), /* show this cap before login */
+	IMAP_CAP_VISIBILITY_POSTAUTH        = BIT(1), /* show this cap after login */
+	IMAP_CAP_VISIBILITY_TLS_ACTIVE      = BIT(2), /* show this cap if tls is active */
+	IMAP_CAP_VISIBILITY_TLS_INACTIVE    = BIT(3), /* show this cap if tls is inactive */
```

Comment:
These are now confusing. Are these set if SSL/TLS is enabled? Or are these set only when STARTTLS is used? Now imap process uses the former way and imap-login uses the latter way. The comments should be at least updated, and if it means just STARTTLS then rename the capabilities to IMAP_CAP_VISIBILITY_STARTTLS_*

Reply:
The flag was originally accompanied by two other flags: 

     IMAP_CAP_VISIBILITY_SSL_ACTIVE
     IMAP_CAP_VISIBILITY_SSL_INACTIVE

However it was raised in a previous review that the differentiation between SSL and TLS was not appropriate, and to remove the SSL flag. 

So given that decision, it would make sense for the remaining flag `IMAP_CAP_VISIBILITY_TLS_ACTIVE` to refer to either one: IMAPS or STARTTLS.

This flag allows for a session to be differentiated between:
 - Secured for any reason (IMAP_CAP_VISIBILITY_SECURE)
 - Secured specifically with TLS/SSL (IMAP_CAP_VISIBILITY_TLS_ACTIVE)

Should the flag still be renamed if it does not correspond directly to STARTTLS?

Reply:
Then it's okay, but the comments in them could say "SSL/TLS" instead of just "tls". And also in imap-login the code is now:

```
	if (client->starttls)
		flags |= IMAP_CAP_VISIBILITY_TLS_ACTIVE;
```

So it should be instead:

```
	if (client->tls)
		flags |= IMAP_CAP_VISIBILITY_TLS_ACTIVE;
```

Although now I see there's nowadays something called client->ssl_secured .. I'm not really sure why we have a difference between that and client->tls. Have to look into that.

Reply:
If my memory serves from when I was investigating those variables (forgive me if I am mistaken, I am no expert):

 - client->starttls is whether imap-login is using STARTTLS
 - client->tls is whether imap-login is using any SSL/TLS/STARTTLS
 - client->ssl_secured is the flag that is passed along to the post-login binary (imap) to indicate whether SSL is active or not (in any form). I believe the post-login (imap) binary has no way of determining whether STARTTLS was used or not -- only whether SSL is effectively active.

Reply:
client->ssl_secured also exists in imap-login. That seems to be used mainly by haproxy support .. I think we'll need to clean up a bunch of this code. Anyway I think client->tls is what you should be using here.